browser/misc_metrics/vertical_tab_metrics.cc

```diff
@@ -0,0 +1,189 @@
+      break;
+    case TabCountType::kGroup:
+    case TabCountType::kPinned:
+      p3a_utils::RecordToHistogramBucket(histogram_name,
+                                         kGroupAndPinnedTabsBuckets, max_value);
+      break;
+  }
+}
+
+}  // namespace
+
+const char kVerticalOpenTabsHistogramName[] = "Brave.VerticalTabs.OpenTabs";
+const char kVerticalGroupTabsHistogramName[] = "Brave.VerticalTabs.GroupTabs";
+const char kVerticalPinnedTabsHistogramName[] = "Brave.VerticalTabs.PinnedTabs";
+
+VerticalTabBrowserMetrics::VerticalTabBrowserMetrics(
+    PrefService* profile_prefs,
+    base::RepeatingClosure change_callback)
+    : profile_prefs_(profile_prefs), change_callback_(change_callback) {
+  for (TabCountType count_type : kAllTabCountTypes) {
+    counts_[count_type] = 0;
+  }
+  pref_change_registrar_.Init(profile_prefs);
+  pref_change_registrar_.Add(
+      brave_tabs::kVerticalTabsEnabled,
+      base::BindRepeating(&VerticalTabBrowserMetrics::UpdateEnabledStatus,
+                          base::Unretained(this)));
+  UpdateEnabledStatus();
+}
+
+VerticalTabBrowserMetrics::~VerticalTabBrowserMetrics() = default;
+
+void VerticalTabBrowserMetrics::OnTabStripModelChanged(
+    TabStripModel* tab_strip_model,
+    const TabStripModelChange& change,
+    const TabStripSelectionChange& selection) {
+  if (!vertical_tabs_enabled_) {
+    return;
+  }
+  counts_[TabCountType::kOpen] = tab_strip_model->count();
+  TabGroupModel* group_model = tab_strip_model->group_model();
+  if (group_model != nullptr) {
+    counts_[TabCountType::kGroup] = group_model->ListTabGroups().size();
+  }
+  counts_[TabCountType::kPinned] = tab_strip_model->IndexOfFirstNonPinnedTab();
+  change_callback_.Run();
+}
+
+bool VerticalTabBrowserMetrics::IsVerticalTabsEnabled() const {
+  return vertical_tabs_enabled_;
+}
+
+size_t VerticalTabBrowserMetrics::GetTabCount(TabCountType count_type) const {
+  return counts_.at(count_type);
+}
+
+void VerticalTabBrowserMetrics::UpdateEnabledStatus() {
+  vertical_tabs_enabled_ =
+      profile_prefs_->GetBoolean(brave_tabs::kVerticalTabsEnabled);
+}
+
+VerticalTabMetrics::VerticalTabMetrics(PrefService* local_state) {
+  for (TabCountType count_type : kAllTabCountTypes) {
+    global_count_storages_[count_type] = std::make_unique<WeeklyStorage>(
+        local_state, GetStoragePrefNameForCountType(count_type));
+  }
+
+  BrowserList::GetInstance()->AddObserver(this);
+}
+
+VerticalTabMetrics::~VerticalTabMetrics() = default;
+
+void VerticalTabMetrics::RegisterPrefs(PrefRegistrySimple* registry) {
+  for (TabCountType count_type : kAllTabCountTypes) {
+    registry->RegisterListPref(GetStoragePrefNameForCountType(count_type));
+  }
+}
+
+void VerticalTabMetrics::UpdateMetrics() {
+  bool is_vertical_tabs_enabled = false;
+  for (const auto& [session_id, browser_metrics] : browser_metrics_) {
+    if (browser_metrics->IsVerticalTabsEnabled()) {
+      is_vertical_tabs_enabled = true;
+    }
```

Comment:
Should we break after is_verticaltabs_enabled_ = true?

Or Can we return right away?

Reply:
Could this be better?
```
  if (base::ranges::any_of(browser_metrics_, [](const auto& id_to_metrics_map){
    return id_to_metrics_map.second->IsVerticalTabsEnabled();
  })) {
    return;
  }
```

Reply:
yes, much better. thanks

Reply:
ended up removing this block entirely since it was redundant. `UpdateMetrics` will only get called by `VerticalTabBrowserMetrics` when vertical tabs is enabled.